### PR TITLE
Improved Github app handling for logs

### PR DESCRIPTION
**Pull request by coverbeck**
**Description**
1. Attempts to improve performance of fetching GitHub orgs, which the UI uses to display orgs with logs.
2. Only allows users to view logs for their repos within an org.

***Background***
In GitHub, a user can be added to an org, or they can be added to specific repos within the org, but not the org itself. We tended to blur the two together. You also have your personal user account, which is not an org.

***Performance***
Use RepositoryListFilter.MEMBER to fetch repos. It should only return repos you have been added to. We were using RepositoryListFilter.ALL to fetch repos, which included repos you had access to because you belonged to the org. Fewer repos in the response, should be faster. Get the orgs you belong to separately.

Seemed a bit faster with curl; results not as clear via Java (but not worse).

***Logs viewing***
* Before, if you could view any repo in an org, you could view the logs for all repos in the org. Now you can only see logs for repos you have permissions to.

**Issue**
* [SEAB-4304](https://ucsc-cgl.atlassian.net/browse/SEAB-4304)
* Improved performance for #4845 

Please make sure that you've checked the following before submitting your pull request. Thanks!

- [x] Check that you pass the basic style checks and unit tests by running `mvn clean install`
- [x] Ensure that the PR targets the correct branch. Check the milestone or fix version of the ticket.
- [x] Follow the existing JPA patterns for queries, using named parameters, to avoid SQL injection
- [x] If you are changing dependencies, check the Snyk status check or the dashboard to ensure you are not introducing new high/critical vulnerabilities
- [x] Assume that inputs to the API can be malicious, and sanitize and/or check for Denial of Service type values, e.g., massive sizes
- [x] Do not serve user-uploaded binary images through the Dockstore API
- [x] Ensure that endpoints that only allow privileged access enforce that with the `@RolesAllowed` annotation
- [x] Do not create cookies, although this may change in the future

### CODE REVIEW

File: dockstore-webservice/src/main/java/io/dockstore/webservice/resources/LambdaEventResource.java [hunk 1]
```diff
@@ -59,21 +62,57 @@ public List<LambdaEvent> getLambdaEventsByOrganization(@ApiParam(hidden = true)
...
+    }
 
-        // If the organization isn't the user's github account, check github to see if we have access to the organization.
-        if (!organization.equals(githubToken.getUsername())) {
-            GitHubSourceCodeRepo sourceCodeRepoInterface = (GitHubSourceCodeRepo)SourceCodeRepoFactory.createSourceCodeRepo(githubToken);
-            if (!sourceCodeRepoInterface.isOneOfMyOrganizations(organization)) {
-                throw new CustomWebApplicationException("You do not have access to the GitHub organization '" + organization + "'", HttpStatus.SC_UNAUTHORIZED);
+    /**
+     * Returns an Optional list of the repositories in the organization the user has access to. If
+     * the user is an organziation member hand has access to all repositories in the organization,
```
Comment (svonworl) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/resources/LambdaEventResource.java:77]: typos

**Comment by denis-yuen** [comment]
> You also have your personal user account, which is not an org.

Can you elaborate on this and how the personal user account (which is not an org, but has org like features like github app logs) would be affected by this PR?

File: dockstore-integration-testing/src/test/java/io/dockstore/webservice/WebhookIT.java [hunk 2]
```diff
@@ -345,6 +345,29 @@ public void testGitHubReleaseNoWorkflowOnDockstore() throws Exception {
...
+    @Test
+    public void testLambdaEvents() throws Exception {
+        CommonTestUtilities.cleanStatePrivate2(SUPPORT, false);
+        final io.dockstore.openapi.client.ApiClient webClient = getOpenAPIWebClient(BasicIT.USER_2_USERNAME, testingPostgres);
+        final io.dockstore.openapi.client.api.UsersApi usersApi = new io.dockstore.openapi.client.api.UsersApi(webClient);
+        final LambdaEventsApi lambdaEventsApi = new LambdaEventsApi(webClient);
+        final List<String> userOrganizations = usersApi.getUserOrganizations("github.com");
+        assertTrue(userOrganizations.contains("dockstoretesting")); // Org user is member of
+        assertTrue(userOrganizations.contains("DockstoreTestUser2")); // The GitHub account
+        final String dockstoreTestUser = "DockstoreTestUser";
```
Comment (denis-yuen) [inline dockstore-integration-testing/src/test/java/io/dockstore/webservice/WebhookIT.java:357]: may need coffee, but the test seems to switch from the second test user above to the un-numbered test user here. Intentional?

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitLabSourceCodeRepo.java [hunk 3]
```diff
@@ -280,11 +279,6 @@ public SourceFile getSourceFile(String path, String id, String branch, Descripto
         return null;
     }
 
-    @Override
```
Comment (denis-yuen) [inline dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitLabSourceCodeRepo.java:283]: Will deleting this and the bitbucket counterpart cause app logs to show up for these two source control methods?

File: dockstore-integration-testing/src/test/java/io/dockstore/webservice/WebhookIT.java (hunk 2 above)
Comment (coverbeck) [reply dockstore-integration-testing/src/test/java/io/dockstore/webservice/WebhookIT.java:357]: Yes, this is intentional, the user `DockstoreTestUser2` has access to a repo in the account of `DockstoreTestuser` (see comment at end of next line).

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java [hunk 4]
```diff
@@ -282,62 +285,80 @@ private Pair<GHContent, String> getContentAndMetadataForFileName(String fileName
...
             return this.handleGetWorkflowGitUrl2RepositoryIdError(e);
         }
     }
 
     /**
-     * Get a list of all the orgs a user has access to
-     * Based on the ALL repository filter, since getAllOrganizations() does not return personal user accounts.
+     * Returns the set of organizations as well as the personal account the user has some level of
+     * access to. Overrides the base implementation for performance, avoiding use of
+     * RepositoryListFilter.MEMBER -- although performance tests results are mixed.
```
[the remark this answers is not in the document]
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitHubSourceCodeRepo.java:328]: It is inaccurate, good catch.

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitLabSourceCodeRepo.java (hunk 3 above)
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitLabSourceCodeRepo.java:283]: The methods weren't and aren't invoked for non-GitHub source controls: [LambdaEventResource](https://github.com/dockstore/dockstore/blob/4e437e6e3bdaba164842c87a49283496bfaa87d9/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/LambdaEventResource.java#L67). And the code does an [explicit cast](https://github.com/dockstore/dockstore/blob/4e437e6e3bdaba164842c87a49283496bfaa87d9/dockstore-webservice/src/main/java/io/dockstore/webservice/resources/LambdaEventResource.java#L86) to GitHubSourceRepo.

I think it's OK, but I'll ponder some more.

**Comment by coverbeck** [comment]
> > You also have your personal user account, which is not an org.
> 
> Can you elaborate on this and how the personal user account (which is not an org, but has org like features like github app logs) would be affected by this PR?

The personal user account should not be affected by this PR in terms of what the end user sees. In terms of implementation:

* Before -- fetch all of the user's repos the user has access to. For me that includes 55 repos in my personal account (I've forked too much!), then derive all the set of organizations from those repos.
* With this PR -- fetch all the user's repos that aren't in their personal account nor in orgs they belong to. Derive the set of orgs from those repos, add the user's personal account, and  the orgs they belong too. It's faster that way in curl, at least for my set of repos.

The `before` description above is simplified; Steve had optimized the lambda endpoint that if the events requested were from your user org, it would short-circuit and say you are authorized (it still does that).

File: dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitLabSourceCodeRepo.java (hunk 3 above)
Comment (coverbeck) [reply dockstore-webservice/src/main/java/io/dockstore/webservice/helpers/GitLabSourceCodeRepo.java:283]: * `io.dockstore.webservice.helpers.SourceCodeRepoInterface#getOrganizations` was previously only called from `io.dockstore.webservice.resources.UserResource#getMyGitHubOrgs`, and only if there was a GitHub token. As part of this PR I have removed `getMyGitHubOrgs`.
* `io.dockstore.webservice.helpers.SourceCodeRepoInterface#getOrganizations` is now more generic, and has a default implementation.

TLDR; I feel this is good.

**Review by svonworl — APPROVED** [review]

**Review by denis-yuen — APPROVED** [review]

**Review by kathy-t — APPROVED** [review]

**Pull request merged**